In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import os

dataset_path = '/content/drive/MyDrive/Deep SAR (SOS) Dataset/'

# Check if the directory exists
if os.path.exists(dataset_path):
    print(f"Listing all contents of: {dataset_path}")
    for root, dirs, files in os.walk(dataset_path):
        # Print the current directory
        print(f"Directory: {root}")
        # Print subdirectories
        if dirs:
            print(f"  Subdirectories: {dirs}")
        # Print files
        if files:
            print(f"  Files: {files}")
        # Removed the break to show all nested files/dirs
else:
    print(f"Directory not found: {dataset_path}. Please ensure it exists and Drive is mounted correctly.")

Listing all contents of: /content/drive/MyDrive/Deep SAR (SOS) Dataset/
Directory: /content/drive/MyDrive/Deep SAR (SOS) Dataset/
  Subdirectories: ['dataset']
Directory: /content/drive/MyDrive/Deep SAR (SOS) Dataset/dataset
  Subdirectories: ['images', 'masks']
Directory: /content/drive/MyDrive/Deep SAR (SOS) Dataset/dataset/images
  Subdirectories: ['val', 'train']
  Files: ['.DS_Store']
Directory: /content/drive/MyDrive/Deep SAR (SOS) Dataset/dataset/images/val
  Files: ['sentinel_11.png', 'palsar_485.png', 'palsar_532.png', 'palsar_308.png', 'sentinel_313.png', 'palsar_693.png', 'palsar_240.png', 'palsar_268.png', 'palsar_320.png', 'sentinel_307.png', 'palsar_20.png', 'palsar_686.png', 'palsar_137.png', 'palsar_651.png', 'palsar_678.png', 'palsar_35.png', 'palsar_644.png', 'sentinel_138.png', 'palsar_123.png', 'sentinel_110.png', 'palsar_136.png', 'palsar_34.png', 'palsar_679.png', 'palsar_645.png', 'sentinel_105.png', 'palsar_692.png', 'palsar_122.png', 'palsar_650.png', 'sentinel

### Dataset Structure Analysis

Based on the output of the `os.walk` command, the dataset is organized in the following hierarchical structure:

*   **`/content/drive/MyDrive/Deep SAR (SOS) Dataset/`** (Top-level directory)
    *   **`dataset/`** (Main dataset folder)
        *   **`images/`** (Contains image files)
            *   **`val/`** (Contains validation images, e.g., `sentinel_11.png`, `palsar_485.png`)
            *   **`train/`** (Contains training images, e.g., `palsar_2049.png`, `sentinel_2582.png`)
        *   **`masks/`** (Likely contains corresponding mask files for the images, though files are not listed in the current output for this directory, the structure implies their presence)
            *   **`val/`**
            *   **`train/`**

This structure clearly separates the image data from its corresponding masks, and further divides both into training and validation sets, which is typical for machine learning tasks involving image segmentation or classification.

In [1]:
import os

dataset_path = '/content/drive/MyDrive/Deep SAR (SOS) Dataset/'
masks_path = os.path.join(dataset_path, 'dataset', 'masks')

# Check if the directory exists
if os.path.exists(masks_path):
    print(f"Listing all contents of: {masks_path}")
    for root, dirs, files in os.walk(masks_path):
        # Print the current directory
        print(f"Directory: {root}")
        # Print subdirectories
        if dirs:
            print(f"  Subdirectories: {dirs}")
        # Print files
        if files:
            print(f"  Files: {files}")
else:
    print(f"Directory not found: {masks_path}. Please ensure it exists and Drive is mounted correctly.")

Listing all contents of: /content/drive/MyDrive/Deep SAR (SOS) Dataset/dataset/masks
Directory: /content/drive/MyDrive/Deep SAR (SOS) Dataset/dataset/masks
  Subdirectories: ['__MACOSX', 'masks']
Directory: /content/drive/MyDrive/Deep SAR (SOS) Dataset/dataset/masks/__MACOSX
  Subdirectories: ['masks']
  Files: ['._masks']
Directory: /content/drive/MyDrive/Deep SAR (SOS) Dataset/dataset/masks/__MACOSX/masks
  Subdirectories: ['val', 'train']
  Files: ['._val', '._.DS_Store', '._train']
Directory: /content/drive/MyDrive/Deep SAR (SOS) Dataset/dataset/masks/__MACOSX/masks/val
  Files: ['._sentinel_705.png', '._sentinel_275.png', '._sentinel_261.png', '._palsar_722.png', '._sentinel_711.png', '._sentinel_17.png', '._sentinel_739.png', '._sentinel_715.png', '._sentinel_513.png', '._palsar_246.png', '._palsar_281.png', '._sentinel_507.png', '._palsar_291.png', '._palsar_252.png', '._sentinel_329.png', '._palsar_1.png', '._sentinel_103.png', '._palsar_326.png', '._sentinel_665.png', '._palsa

**Reasoning**:
To analyze the file alignment between images and masks for both training and validation sets, I need to extract the filenames from the relevant directories. I will iterate through the `images` and `masks` directories to collect all filenames, excluding any metadata files. Then, I will compare the file counts and check for exact filename matches for each split.



In [2]:
import os

dataset_path = '/content/drive/MyDrive/Deep SAR (SOS) Dataset/'

images_train_path = os.path.join(dataset_path, 'dataset', 'images', 'train')
images_val_path = os.path.join(dataset_path, 'dataset', 'images', 'val')
masks_train_path = os.path.join(dataset_path, 'dataset', 'masks', 'masks', 'train')
masks_val_path = os.path.join(dataset_path, 'dataset', 'masks', 'masks', 'val')


def get_filenames(path, strip_prefix=None):
    filenames = set()
    if os.path.exists(path):
        for root, _, files in os.walk(path):
            for file in files:
                # Exclude macOS metadata files like .DS_Store or files starting with ._
                if not file.startswith('.') and file.endswith('.png'):
                    if strip_prefix and file.startswith(strip_prefix):
                        filenames.add(file[len(strip_prefix):])
                    else:
                        filenames.add(file)
    return filenames

# Get filenames for images
images_train_files = get_filenames(images_train_path)
images_val_files = get_filenames(images_val_path)

# Get filenames for masks, stripping the '._' prefix from __MACOSX files if present
# From the os.walk output, files in masks/__MACOSX/masks/train/ and masks/__MACOSX/masks/val/
# had '._' prefix. The main masks/masks/train and masks/masks/val directly contain files without the prefix.
# Need to adjust logic to fetch files from the correct masks path based on the previous output

# Re-inspecting the os.walk output for masks, the actual files are in
# Directory: /content/drive/MyDrive/Deep SAR (SOS) Dataset/dataset/masks/masks/val
# Directory: /content/drive/MyDrive/Deep SAR (SOS) Dataset/dataset/masks/masks/train
# these do NOT have the ._ prefix, the ._ files are in __MACOSX directories which are invalid.

masks_train_files = get_filenames(masks_train_path)
masks_val_files = get_filenames(masks_val_path)


# Compare counts
print(f"--- File Counts ---")
print(f"Images Train: {len(images_train_files)}")
print(f"Masks Train:  {len(masks_train_files)}")
print(f"Images Val:   {len(images_val_files)}")
print(f"Masks Val:    {len(masks_val_files)}")

# Compare file names for training set
missing_in_masks_train = images_train_files - masks_train_files
missing_in_images_train = masks_train_files - images_train_files

print(f"\n--- Training Set Alignment ---")
print(f"Files in images/train but not in masks/train: {len(missing_in_masks_train)}")
print(f"Files in masks/train but not in images/train: {len(missing_in_images_train)}")

# Compare file names for validation set
missing_in_masks_val = images_val_files - masks_val_files
missing_in_images_val = masks_val_files - images_val_files

print(f"\n--- Validation Set Alignment ---")
print(f"Files in images/val but not in masks/val: {len(missing_in_masks_val)}")
print(f"Files in masks/val but not in images/val: {len(missing_in_images_val)}")

# Check for specific missing files if any (show first few if many)
if missing_in_masks_train: print(f"  Examples missing in masks/train: {list(missing_in_masks_train)[:5]}")
if missing_in_images_train: print(f"  Examples missing in images/train: {list(missing_in_images_train)[:5]}")
if missing_in_masks_val: print(f"  Examples missing in masks/val: {list(missing_in_masks_val)[:5]}")
if missing_in_images_val: print(f"  Examples missing in images/val: {list(missing_in_images_val)[:5]}")

--- File Counts ---
Images Train: 6455
Masks Train:  6459
Images Val:   1389
Masks Val:    1615

--- Training Set Alignment ---
Files in images/train but not in masks/train: 0
Files in masks/train but not in images/train: 4

--- Validation Set Alignment ---
Files in images/val but not in masks/val: 0
Files in masks/val but not in images/val: 226
  Examples missing in images/train: ['palsar_1392 (1).png', 'palsar_1393 (1).png', 'palsar_1345 (1).png', 'palsar_135 (1).png']
  Examples missing in images/val: ['sentinel_711.png', 'sentinel_74.png', 'sentinel_705.png', 'sentinel_88.png', 'sentinel_690.png']


### Dataset Alignment Summary

Based on the comparison of filenames between the `images` and `masks` directories for both training and validation sets, the following observations were made:

*   **Training Set Alignment**:
    *   All `images/train` files have a corresponding mask in `masks/train` (0 missing in masks).
    *   However, 4 mask files in `masks/train` do not have a corresponding image file in `images/train`. The example names `['palsar_1392 (1).png', 'palsar_1393 (1).png', 'palsar_1345 (1).png', 'palsar_135 (1).png']` suggest these might be duplicates or incorrectly named files within the mask folder that include a `(1)` suffix. These files should ideally be removed from the mask folder or their corresponding images renamed if they exist.

*   **Validation Set Alignment**:
    *   All `images/val` files have a corresponding mask in `masks/val` (0 missing in masks).
    *   A significant discrepancy was found in the validation set: 226 mask files in `masks/val` do not have a corresponding image file in `images/val`. Examples like `['sentinel_711.png', 'sentinel_74.png', 'sentinel_705.png', 'sentinel_88.png', 'sentinel_690.png']` indicate that these are regular filenames, not necessarily duplicates with a special suffix.

**Conclusion and Next Steps**:

While the dataset guarantees that every image has a mask, there are mask files without corresponding images. These misaligned mask files could lead to errors during model training or evaluation if not handled properly. To ensure data integrity for an image segmentation task, it is crucial that every image has an exactly corresponding mask, and vice-versa. Therefore, the mask files that do not have a corresponding image file should be identified and either removed from the dataset or their missing image counterparts should be located and added.

For the next steps, it is recommended to:
1.  **Remove or address the 4 misaligned mask files in the training set**.
2.  **Identify and remove the 226 misaligned mask files in the validation set** to maintain a consistent 1:1 mapping between images and masks.

## Clean up misaligned mask files

### Subtask:
Remove mask files from the training and validation sets that do not have corresponding image files.


**Reasoning**:
I will construct the full paths for the misaligned mask files using the previously identified sets and paths, then remove these files using `os.remove()` to clean up the dataset.



In [3]:
import os

# Construct full paths and remove missing mask files from the training set
print("Removing misaligned mask files from training set...")
for filename in missing_in_images_train:
    file_path = os.path.join(masks_train_path, filename)
    if os.path.exists(file_path):
        os.remove(file_path)
        print(f"Removed: {file_path}")
    else:
        print(f"Warning: {file_path} not found for removal.")

# Construct full paths and remove missing mask files from the validation set
print("\nRemoving misaligned mask files from validation set...")
for filename in missing_in_images_val:
    file_path = os.path.join(masks_val_path, filename)
    if os.path.exists(file_path):
        os.remove(file_path)
        print(f"Removed: {file_path}")
    else:
        print(f"Warning: {file_path} not found for removal.")

# Re-run the file count and alignment check to confirm
# The get_filenames function is already defined in the previous cell

images_train_files = get_filenames(images_train_path)
images_val_files = get_filenames(images_val_path)
masks_train_files = get_filenames(masks_train_path)
masks_val_files = get_filenames(masks_val_path)

# Compare counts
print(f"\n--- Re-checked File Counts ---")
print(f"Images Train: {len(images_train_files)}")
print(f"Masks Train:  {len(masks_train_files)}")
print(f"Images Val:   {len(images_val_files)}")
print(f"Masks Val:    {len(masks_val_files)}")

# Compare file names for training set
missing_in_masks_train_recheck = images_train_files - masks_train_files
missing_in_images_train_recheck = masks_train_files - images_train_files

print(f"\n--- Re-checked Training Set Alignment ---")
print(f"Files in images/train but not in masks/train: {len(missing_in_masks_train_recheck)}")
print(f"Files in masks/train but not in images/train: {len(missing_in_images_train_recheck)}")

# Compare file names for validation set
missing_in_masks_val_recheck = images_val_files - masks_val_files
missing_in_images_val_recheck = masks_val_files - images_val_files

print(f"\n--- Re-checked Validation Set Alignment ---")
print(f"Files in images/val but not in masks/val: {len(missing_in_masks_val_recheck)}")
print(f"Files in masks/val but not in images/val: {len(missing_in_images_val_recheck)}")

# Confirm both are empty
if not missing_in_images_train_recheck and not missing_in_images_val_recheck:
    print("\nAll misaligned mask files successfully removed. Training and validation sets are now aligned.")
else:
    print("\nSome misaligned mask files still exist or new discrepancies found. Further investigation may be needed.")

Removing misaligned mask files from training set...
Removed: /content/drive/MyDrive/Deep SAR (SOS) Dataset/dataset/masks/masks/train/palsar_1392 (1).png
Removed: /content/drive/MyDrive/Deep SAR (SOS) Dataset/dataset/masks/masks/train/palsar_1393 (1).png
Removed: /content/drive/MyDrive/Deep SAR (SOS) Dataset/dataset/masks/masks/train/palsar_1345 (1).png
Removed: /content/drive/MyDrive/Deep SAR (SOS) Dataset/dataset/masks/masks/train/palsar_135 (1).png

Removing misaligned mask files from validation set...
Removed: /content/drive/MyDrive/Deep SAR (SOS) Dataset/dataset/masks/masks/val/sentinel_711.png
Removed: /content/drive/MyDrive/Deep SAR (SOS) Dataset/dataset/masks/masks/val/sentinel_74.png
Removed: /content/drive/MyDrive/Deep SAR (SOS) Dataset/dataset/masks/masks/val/sentinel_705.png
Removed: /content/drive/MyDrive/Deep SAR (SOS) Dataset/dataset/masks/masks/val/sentinel_88.png
Removed: /content/drive/MyDrive/Deep SAR (SOS) Dataset/dataset/masks/masks/val/sentinel_690.png
Removed: /co

## Organize Dataset into Structured Format

### Subtask:
Create a test dataset by splitting a portion of the existing training data (images and masks) into a new test set, maintaining the current validation set.


**Reasoning**:
The first step is to define all necessary file paths for the new training, validation, and test sets, including creating the new test directories to store the split data.



In [4]:
import os
import shutil
import random

# Define paths for the new dataset structure
# Original training paths (these will become the new training paths after moving files)
new_images_train_path = images_train_path
new_masks_train_path = masks_train_path

# Validation paths remain the same
new_images_val_path = images_val_path
new_masks_val_path = masks_val_path

# New test paths
test_images_path = os.path.join(dataset_path, 'dataset', 'images', 'test')
test_masks_path = os.path.join(dataset_path, 'dataset', 'masks', 'test')

# Create new test directories if they don't exist
os.makedirs(test_images_path, exist_ok=True)
os.makedirs(test_masks_path, exist_ok=True)

print(f"New training images path: {new_images_train_path}")
print(f"New training masks path: {new_masks_train_path}")
print(f"New validation images path: {new_images_val_path}")
print(f"New validation masks path: {new_masks_val_path}")
print(f"New test images path: {test_images_path}")
print(f"New test masks path: {test_masks_path}")

print("Created test directories if they did not exist.")

New training images path: /content/drive/MyDrive/Deep SAR (SOS) Dataset/dataset/images/train
New training masks path: /content/drive/MyDrive/Deep SAR (SOS) Dataset/dataset/masks/masks/train
New validation images path: /content/drive/MyDrive/Deep SAR (SOS) Dataset/dataset/images/val
New validation masks path: /content/drive/MyDrive/Deep SAR (SOS) Dataset/dataset/masks/masks/val
New test images path: /content/drive/MyDrive/Deep SAR (SOS) Dataset/dataset/images/test
New test masks path: /content/drive/MyDrive/Deep SAR (SOS) Dataset/dataset/masks/test
Created test directories if they did not exist.


**Reasoning**:
Now that the paths are defined and test directories created, the next step is to determine the split ratio, collect filenames from the original training set, and then randomly split these filenames into new training and test sets while ensuring corresponding image and mask files stay together.



In [5]:
import random
import shutil

# Define split ratio (e.g., 20% of original training data for the new test set)
TEST_SPLIT_RATIO = 0.2

# Get all image filenames from the original training set
original_train_images = sorted(list(get_filenames(new_images_train_path)))

# Ensure masks are aligned before splitting (this was handled in the previous step)
# We'll rely on the fact that all original_train_images now have corresponding masks

# Determine the number of files for the new test set
num_total_train_files = len(original_train_images)
num_test_files = int(num_total_train_files * TEST_SPLIT_RATIO)

# Randomly select filenames for the new test set
random.seed(42) # for reproducibility
test_filenames = random.sample(original_train_images, num_test_files)

# The remaining filenames will form the new training set
new_train_filenames = [f for f in original_train_images if f not in test_filenames]

print(f"Original training set size: {num_total_train_files}")
print(f"Number of files moved to new test set: {len(test_filenames)}")
print(f"New training set size: {len(new_train_filenames)}")

# Move selected files to the new test directories
print("\nMoving files to new test directories...")
for filename in test_filenames:
    # Move image file
    shutil.move(os.path.join(new_images_train_path, filename),
                os.path.join(test_images_path, filename))
    # Move mask file
    shutil.move(os.path.join(new_masks_train_path, filename),
                os.path.join(test_masks_path, filename))
print("Finished moving files.")

Original training set size: 6455
Number of files moved to new test set: 1291
New training set size: 5164

Moving files to new test directories...
Finished moving files.


**Reasoning**:
Now that the files have been moved to create the new test set, it's crucial to verify the file counts and alignment across the new training, validation, and test sets. This ensures that the splitting and moving operations were successful and that all sets maintain a 1:1 image-to-mask correspondence, confirming the dataset is ready for model training and evaluation.



In [6]:
import os

# Re-run get_filenames for all sets to get updated counts and file lists
# The get_filenames function was defined in a previous cell.

# Get filenames for new training set
new_images_train_files = get_filenames(new_images_train_path)
new_masks_train_files = get_filenames(new_masks_train_path)

# Get filenames for validation set (should remain unchanged from previous check)
new_images_val_files = get_filenames(new_images_val_path)
new_masks_val_files = get_filenames(new_masks_val_path)

# Get filenames for new test set
test_images_files = get_filenames(test_images_path)
test_masks_files = get_filenames(test_masks_path)

# Compare counts for all sets
print(f"\n--- Final File Counts ---")
print(f"New Images Train: {len(new_images_train_files)}")
print(f"New Masks Train:  {len(new_masks_train_files)}")
print(f"Validation Images: {len(new_images_val_files)}")
print(f"Validation Masks:  {len(new_masks_val_files)}")
print(f"Test Images:      {len(test_images_files)}")
print(f"Test Masks:       {len(test_masks_files)}")

# Verify alignment for new training set
missing_in_masks_new_train = new_images_train_files - new_masks_train_files
missing_in_images_new_train = new_masks_train_files - new_images_train_files
print(f"\n--- New Training Set Alignment ---")
print(f"Files in new images/train but not in new masks/train: {len(missing_in_masks_new_train)}")
print(f"Files in new masks/train but not in new images/train: {len(missing_in_images_new_train)}")

# Verify alignment for validation set
missing_in_masks_val_final = new_images_val_files - new_masks_val_files
missing_in_images_val_final = new_masks_val_files - new_images_val_files
print(f"\n--- Validation Set Alignment (Final) ---")
print(f"Files in images/val but not in masks/val: {len(missing_in_masks_val_final)}")
print(f"Files in masks/val but not in images/val: {len(missing_in_images_val_final)}")

# Verify alignment for new test set
missing_in_masks_test = test_images_files - test_masks_files
missing_in_images_test = test_masks_files - test_images_files
print(f"\n--- New Test Set Alignment ---")
print(f"Files in test_images_path but not in test_masks_path: {len(missing_in_masks_test)}")
print(f"Files in test_masks_path but not in test_images_path: {len(missing_in_images_test)}")

# Final confirmation
if not missing_in_masks_new_train and not missing_in_images_new_train and \
   not missing_in_masks_val_final and not missing_in_images_val_final and \
   not missing_in_masks_test and not missing_in_images_test:
    print("\nAll image and mask sets (new train, val, test) are now perfectly aligned.")
else:
    print("\nDiscrepancies found after splitting. Further investigation is needed.")


--- Final File Counts ---
New Images Train: 5164
New Masks Train:  5164
Validation Images: 1389
Validation Masks:  1389
Test Images:      1291
Test Masks:       1291

--- New Training Set Alignment ---
Files in new images/train but not in new masks/train: 0
Files in new masks/train but not in new images/train: 0

--- Validation Set Alignment (Final) ---
Files in images/val but not in masks/val: 0
Files in masks/val but not in images/val: 0

--- New Test Set Alignment ---
Files in test_images_path but not in test_masks_path: 0
Files in test_masks_path but not in test_images_path: 0

All image and mask sets (new train, val, test) are now perfectly aligned.


## Summary:

### Q&A
*   **Determine if a separate test set needs to be created or extracted from the existing data.**
    Yes, a separate test set was deemed necessary and was successfully created by splitting a portion of the existing training data.

### Data Analysis Key Findings
*   The initial dataset required manual acquisition from Kaggle, uploading to Google Drive, and mounting in Google Colab.
*   Initial inspection of the dataset revealed misalignments between image and mask files:
    *   The training set contained 6455 image files but 6459 mask files, with 4 mask files lacking corresponding images.
    *   The validation set contained 1389 image files but 1615 mask files, with 226 mask files lacking corresponding images.
*   Misaligned mask files were successfully identified and removed from both sets: 4 from the training set and 226 from the validation set.
*   After cleanup, both training and validation sets achieved perfect alignment, ensuring a 1:1 correspondence between images and masks:
    *   Training set: 6455 image files and 6455 mask files.
    *   Validation set: 1389 image files and 1389 mask files.
*   A new test set was created by splitting 20% of the *cleaned* training data (1291 image-mask pairs) and moving them to dedicated test directories.
*   The final dataset structure for model training and evaluation is as follows, with perfect alignment (1:1 image to mask ratio) in each split:
    *   Training set: 5164 image-mask pairs.
    *   Validation set: 1389 image-mask pairs.
    *   Test set: 1291 image-mask pairs.

### Insights or Next Steps
*   The dataset is now meticulously clean and organized into well-aligned training, validation, and test sets, ready for the development and robust evaluation of an oil spill detection segmentation model.
*   The established dataset structure with distinct splits minimizes the risk of data leakage and provides a reliable framework for assessing model generalization performance.
